docs/source/essentials/normalized-cache.mdx

```diff
@@ -0,0 +1,350 @@
+  buildTypes {
+    custom {
+      // your code...
+      matchingFallbacks = ["debug"]
+    }
+  }
+}
+```
+
+</MultiCodeBlock>
+
+Once the dependency is added, create the `SqlNormalizedCacheFactory`:
+
+<MultiCodeBlock>
+
+```kotlin
+// Android
+val sqlNormalizedCacheFactory = SqlNormalizedCacheFactory(context, "apollo.db")
+// JVM
+val sqlNormalizedCacheFactory = SqlNormalizedCacheFactory("jdbc:sqlite:apollo.db")
+// iOS
+val sqlNormalizedCacheFactory = SqlNormalizedCacheFactory("apollo.db")
+
+// Build the ApolloClient
+val apolloClient = ApolloClient.builder()
+  .serverUrl("https://...")
+  .normalizedCache(sqlNormalizedCacheFactory)
+  .build())
+```
+
+```java
+// Android
+SqlNormalizedCacheFactory sqlNormalizedCacheFactory = new SqlNormalizedCacheFactory(context, "apollo.db")
+// JVM
+SqlNormalizedCacheFactory sqlNormalizedCacheFactory = new SqlNormalizedCacheFactory("jdbc:sqlite:apollo.db")
+
+// Build the ApolloClient
+ApolloClient apolloClient = ApolloClient.builder()
+  .serverUrl("https://...")
+  .normalizedCache(sqlNormalizedCacheFactory)
+  .build();
+```
+
+</MultiCodeBlock>
+
+## Chaining caches
+
+To get the best of both caches, you can chain a `LruNormalizedCacheFactory` with a `SqlNormalizedCacheFactory`:
+
+<MultiCodeBlock>
+
+```kotlin:title=Kotlin
+
+val sqlCacheFactory = SqlNormalizedCacheFactory(context, "db_name")
+val memoryFirstThenSqlCacheFactory = LruNormalizedCacheFactory(
+    EvictionPolicy.builder().maxSizeBytes(10 * 1024).build()
+).chain(sqlCacheFactory)
+
+```
+
+```java:title=Java
+
+NormalizedCacheFactory sqlCacheFactory = new SqlNormalizedCacheFactory(context, "db_name");
+NormalizedCacheFactory memoryFirstThenSqlCacheFactory = new LruNormalizedCacheFactory(
+  EvictionPolicy.builder().maxSizeBytes(10 * 1024).build()
+).chain(sqlCacheFactory);
+
+```
+</MultiCodeBlock>
+
+Reads will read from the first cache hit in the chain. Writes will propagate down the entire chain.
+
+## Specifying your object IDs
+
+By default, Apollo Android uses the field path as key to store data. Back to the original example:
+
+```graphl
+query BookWithAuthorName {
+  favoriteBook {
+    id
+    title
+    author {
+      id
+      name
+    }
+  }
+}
+
+query AuthorById($id: String!) {
+  author(id: $id) {
+      id
+      name
+    }
+  }
+}
+```
+
+This will store the following records:
+
+* `"favoriteBook"`: `{"id": "book1", "title": "Les guerriers du silence", "author": "ApolloCacheReference{favoriteBook.author}"}`
+* `"favoriteBook.author"`: `{"id": "author1", name": "Pierre Bordage"}`
+* `"author("id": "author1")"`: `{"id": "author1", "name": "Pierre Bordage"}`
+* `"QUERY_ROOT"`: `{"favoriteBook": "ApolloCacheReference{favoriteBook}", "author(\"id\": \"author1\")": "ApolloCacheReference{author(\"id\": \"author1\")}"}`
+
+As you can see, the author is stored twice. This is unfortunate because it takes more space but also because changing data in one place will not notify other watchers. What we want instead is this:
```

Comment:
"This is undesirable, both because it takes more space, and because modifying one of those objects will not notify the watchers of the other."